xdsl/parser.py

```diff
@@ -829,11 +829,10 @@ def try_parse_value_id_and_type(self) -> tuple[Span, Attribute] | None:
             return value_id, type
 
     def try_parse_type(self) -> Attribute | None:
-        if (builtin_type := self.try_parse_builtin_type()) is not None:
-            return builtin_type
-        if (dialect_type := self.try_parse_dialect_type()) is not None:
-            return dialect_type
-        return None
+        if not self.tokenizer.starts_with('!'):
```

Comment:
This was to avoid some spurious errors in a malformed filecheck. Dialect types must start with ! anyway, so this avoids an extra backtracking

Reply:
Behaviour is not changed

Reply:
I think I would prefer to remove the `not` and switch the branches around. That removes a negation which makes it easier to read.

Reply:
Agreed